src/components/CountDown/CountDown.tsx

```diff
@@ -0,0 +1,133 @@
+import React, {useEffect, useRef, useState} from 'react'
+import {AppState, type StyleProp, type ViewStyle, type TextStyle} from 'react-native'
+import styled from 'styled-components/native'
+
+export type CountDownProps = {
+  /*
+  init time countdown by second
+  */
+  initialSeconds: number
+  /*
+  call when count down finish
+  */
+  onFinish?: () => void
+  /*
+  container countdown view style
+  */
+  containerStyle?: StyleProp<ViewStyle>
+  /*
+  text countdown style
+  */
+  textStyle?: StyleProp<TextStyle>
+  /*
+  loop countdown
+  */
+  loop?: boolean
+  /*
+  format countdown
+  */
+  format?: 'MM:SS' | 'HH:MM:SS' | 'DD:HH:MM:SS'
+}
+const intervalTimeBySecond = 1000
+const CountDown: React.FunctionComponent<CountDownProps> = ({
+  initialSeconds,
+  containerStyle,
+  onFinish,
+  textStyle,
+  loop,
+  format = 'MM:SS',
+}) => {
+  const [seconds, setSeconds] = useState(initialSeconds)
+
+  const appState = useRef(AppState.currentState)
+  const timeEnd = useRef(new Date(new Date().getTime() + initialSeconds * 1000).getTime())
+  const [appStateVisible, setAppStateVisible] = useState(appState.current)
+
+  /*
+    check app state is change
+  */
+  useEffect(() => {
+    const subscription = AppState.addEventListener('change', nextAppState => {
+      if (appState.current.match(/inactive|background/) && nextAppState === 'active') {
+        console.log('App has come to the foreground!')
+      }
+      appState.current = nextAppState
+      setAppStateVisible(appState.current)
+    })
+    return () => {
+      subscription.remove()
+    }
+  }, [])
+
+  useEffect(() => {
+    const interval = setInterval(() => {
+      if (seconds >= 0) {
+        const uff = new Date(timeEnd.current - new Date().getTime()).getTime()
+        setSeconds(uff / 1000)
+      }
+      if (seconds < 1) {
+        if (loop) {
+          timeEnd.current = new Date(new Date().getTime() + initialSeconds * 1000).getTime()
+          setSeconds(initialSeconds)
+        } else {
+          clearInterval(interval)
+          if (onFinish) {
+            onFinish()
+          }
+        }
+      }
+    }, intervalTimeBySecond)
+    return () => {
+      clearInterval(interval)
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [seconds, appStateVisible])
```

Comment:
This `setInterval` is change every second we setSeconds, maybe we should just use `setTimeout`, that way we won't need to clearInterval

Reply:
i will check it